// marko/dist/runtime/components/ServerComponent.js
"use strict";


class ServerComponent {
  constructor(id, input, out, typeName, customEvents, scope) {
    this.id = id;
    this.O_ = customEvents;
    this.G_ = scope;
    this.typeName = typeName;
    this.N_ = undefined; // Used to keep track of bubbling DOM events for components rendered on the server
    this._W_ = 0;

    this.onCreate(input, out);
    this._X_ = this.onInput(input, out) || input;
    if (this.Q_ === undefined) {
      this.Q_ = this._X_;
    }
    this.onRender(out);
  }

  set input(newInput) {
    this.Q_ = newInput;
  }

  get input() {
    return this.Q_;
  }

  set state(newState) {
    this.J_ = newState;
  }

  get state() {
    return this.J_;
  }

  get _t_() {
    return this.J_;
  }

  elId(nestedId) {
    var id = this.id;

    if (nestedId == null) {
      return id;
    } else {
      if (typeof nestedId !== "string") {
        // eslint-disable-next-line no-constant-condition




        nestedId = String(nestedId);
      }

      if (nestedId.indexOf("#") === 0) {
        id = "#" + id;
        nestedId = nestedId.substring(1);
      }

      return id + "-" + nestedId;
    }
  }

  onCreate() {}
  onInput() {}
  onRender() {}}


ServerComponent.prototype.getElId = ServerComponent.prototype.elId;

module.exports = ServerComponent;
