// marko/dist/runtime/RenderResult.js
"use strict";var domInsert = require("./dom-insert");


function getComponentDefs(result) {
  var componentDefs = result.b_;

  if (!componentDefs) {
    throw Error("No component");
  }
  return componentDefs;
}

function RenderResult(out) {
  this.out = this.y_ = out;
  this.b_ = undefined;
}

module.exports = RenderResult;

var proto = RenderResult.prototype = {
  getComponent: function () {
    return this.getComponents()[0];
  },
  getComponents: function (selector) {
    if (this.b_ === undefined) {
      throw Error("Not added to DOM");
    }

    var componentDefs = getComponentDefs(this);

    var components = [];

    componentDefs.forEach(function (componentDef) {
      var component = componentDef.h_;
      if (!selector || selector(component)) {
        components.push(component);
      }
    });

    return components;
  },

  afterInsert: function (doc) {
    var out = this.y_;
    var componentsContext = out.b_;
    if (componentsContext) {
      this.b_ = componentsContext.z_(doc);
    } else {
      this.b_ = null;
    }

    return this;
  },
  getNode: function (doc) {
    return this.y_.A_(doc);
  },
  getOutput: function () {
    return this.y_.B_();
  },
  toString: function () {
    return this.y_.toString();
  },
  document: typeof document != "undefined" && document };


Object.defineProperty(proto, "html", {
  get: function () {
    // eslint-disable-next-line no-constant-condition





    return this.toString();
  } });


Object.defineProperty(proto, "context", {
  get: function () {
    // eslint-disable-next-line no-constant-condition





    return this.y_;
  } });


// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(
proto,
function getEl(renderResult, referenceEl) {
  return renderResult.getNode(referenceEl.ownerDocument);
},
function afterInsert(renderResult, referenceEl) {
  var isShadow =
  typeof ShadowRoot === "function" && referenceEl instanceof ShadowRoot;
  return renderResult.afterInsert(
  isShadow ? referenceEl : referenceEl.ownerDocument);

});
