// @marko/babel-utils/dist/index.js
"use strict";exports.__esModule = true;exports.getTagDefForTagName = exports.getTaglibLookup = exports.importNamed = exports.importDefault = exports.resolveRelativePath = exports.parseExpression = exports.parseScript = exports.withLoc = exports.getLocRange = exports.getLoc = exports.normalizeTemplateString = exports.assertNoAttributeTags = exports.assertNoVar = exports.assertNoParams = exports.assertNoAttributes = exports.assertNoArgs = exports.assertAllowedAttributes = exports.resolveTagImport = exports.getTemplateId = exports.loadFileForImport = exports.loadFileForTag = exports.getArgOrSequence = exports.findAttributeTags = exports.findParentTag = exports.getFullyResolvedTagName = exports.getTagDef = exports.getMacroIdentifier = exports.isLoopTag = exports.isTransparentTag = exports.isAttributeTag = exports.isDynamicTag = exports.isMacroTag = exports.isNativeTag = void 0;var _tags = require("./tags");exports.isNativeTag = _tags.isNativeTag;exports.isMacroTag = _tags.isMacroTag;exports.isDynamicTag = _tags.isDynamicTag;exports.isAttributeTag = _tags.isAttributeTag;exports.isTransparentTag = _tags.isTransparentTag;exports.isLoopTag = _tags.isLoopTag;exports.getMacroIdentifier = _tags.getMacroIdentifier;exports.getTagDef = _tags.getTagDef;exports.getFullyResolvedTagName = _tags.getFullyResolvedTagName;exports.findParentTag = _tags.findParentTag;exports.findAttributeTags = _tags.findAttributeTags;exports.getArgOrSequence = _tags.getArgOrSequence;exports.loadFileForTag = _tags.loadFileForTag;exports.loadFileForImport = _tags.loadFileForImport;exports.getTemplateId = _tags.getTemplateId;exports.resolveTagImport = _tags.resolveTagImport;

















var _assert = require("./assert");exports.assertAllowedAttributes = _assert.assertAllowedAttributes;exports.assertNoArgs = _assert.assertNoArgs;exports.assertNoAttributes = _assert.assertNoAttributes;exports.assertNoParams = _assert.assertNoParams;exports.assertNoVar = _assert.assertNoVar;exports.assertNoAttributeTags = _assert.assertNoAttributeTags;







var _templateString = require("./template-string");exports.normalizeTemplateString = _templateString.normalizeTemplateString;

var _loc = require("./loc");exports.getLoc = _loc.getLoc;exports.getLocRange = _loc.getLocRange;exports.withLoc = _loc.withLoc;

var _parse = require("./parse");exports.parseScript = _parse.parseScript;exports.parseExpression = _parse.parseExpression;

var _imports = require("./imports");exports.resolveRelativePath = _imports.resolveRelativePath;exports.importDefault = _imports.importDefault;exports.importNamed = _imports.importNamed;

var _taglib = require("./taglib");exports.getTaglibLookup = _taglib.getTaglibLookup;exports.getTagDefForTagName = _taglib.getTagDefForTagName;
